### PR TITLE
types: handle io.EOF error when to parse datatime

### LINKED ISSUE #35678: Unexpected Connection Lost
### 1. Minimal reproduce step (Required)
```sql
SELECT CAST(REVERSE(-1502572527) AS DATE); -- ERROR 2013 (HY000): Lost connection to MySQL server during query
```
### 2. What did you expect to see? (Required)
No error.
### 3. What did you see instead (Required)
ERROR 2013 (HY000): Lost connection to MySQL server during query
### 4. What is your TiDB version? (Required)

| Release Version: v6.2.0-alpha-159-ge0527ba27
Edition: Community
Git Commit Hash: e0527ba27c72b0a533b126fedfa025d47a209ca9
Git Branch: master
UTC Build Time: 2022-06-21 11:55:54
GoVersion: go1.18.3
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
Store: unistore |

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #35678

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/executor_issue_test.go [hunk 1]
```diff
@@ -684,6 +684,11 @@ func TestIssue22231(t *testing.T) {
 	tk.MustQuery("select cast('2020-05-28 23:59:59 00:00:00' as datetime)").Check(testkit.Rows("2020-05-28 23:59:59"))
 	tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1292 Truncated incorrect datetime value: '2020-05-28 23:59:59 00:00:00'"))
 	tk.MustExec("drop table if exists t_issue_22231")
+
+	tk.MustQuery("SELECT CAST(\"1111111111-\" AS DATE);")
+	tk.MustQuery("SHOW WARNINGS").
+		Check(
```
Comment (Defined2014) [inline executor/executor_issue_test.go:690]: Maybe combine L689-L691 to one line is better.

**Review by Defined2014 — APPROVED** [review]
others LGTM

File: executor/executor_issue_test.go (hunk 1 above)
Comment (hawkingrei) [reply executor/executor_issue_test.go]: Done

**Review by xhebox — APPROVED** [review]

**Pull request merged**